### PR TITLE
[mlir][LLVMIR] Add operand bundle support for llvm.intr.assume

### PR DESCRIPTION
**Pull request by Lancern**
This PR adds operand bundle support for `llvm.intr.assume`.

This PR actually contains two parts:

- `llvm.intr.assume` now accepts operand bundle related attributes and operands. `llvm.intr.assume` does not take constraint on the operand bundles, but obviously only a few set of operand bundles are meaningful. I plan to add some of those (e.g. `aligned` and `separate_storage` are what interest me but other people may be interested in other operand bundles as well) in future patches.

- The definitions of `llvm.call`, `llvm.invoke`, and `llvm.call_intrinsic` actually define `op_bundle_tags` as an operation property. It turns out this approach would introduce some unnecessary burden if applied equally to the intrinsic operations because properties are not available through `Operation *` but we have to operate on `Operation *` during the import/export of intrinsics, so this PR changes it from a property to an array attribute.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:853]: Similarly `opBundleSizesAttrName` and `opBundleTagsAttrName` could probably be constants?

**Review by gysit — COMMENTED** [review]
Thanks a lot for following up on this!

I added some suggestions around the tablegen modifications. In particular, it could make sense to follow the implementation of the alias attributes. Let me know if that doesn't make sense or if you have questions.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td
Comment (Lancern) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td]: Updated.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOpBase.td
Comment (Lancern) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOpBase.td]: Updated.

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMIRToLLVMTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMIRToLLVMTranslation.cpp]: Updated.

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp]: Updated.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Updated.

File: mlir/test/Dialect/LLVMIR/roundtrip.mlir
Comment (Lancern) [reply mlir/test/Dialect/LLVMIR/roundtrip.mlir]: Updated.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td
Comment (Lancern) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td]: Added `requiresOperandBundles` and removed the two `XXXAttrName` parameters.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td
Comment (Lancern) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td]: Updated.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp]: Added constant getters in `LLVMDialect.td` and updated code accordingly.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp]: Updated.

File: mlir/test/Target/LLVMIR/Import/intrinsic.ll
Comment (Lancern) [reply mlir/test/Target/LLVMIR/Import/intrinsic.ll]: Added a test for export in `mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir`.

**Comment by zero9178** [comment]
> It turns out this approach would introduce some unnecessary burden if applied equally to the intrinsic operations, so this PR changes it from a property to an array attribute.

What are those burdens? Don't really care either way, but I'm just very curious 🙂 Might also be worth summarizing super shortly in the commit message

File: mlir/test/Target/LLVMIR/Import/intrinsic.ll
Comment (gysit) [inline mlir/test/Target/LLVMIR/Import/intrinsic.ll:626]:
```suggestion
; CHECK-SAME:  %[[TRUE:[a-zA-Z0-9]+]]
; CHECK-SAME:  %[[PTR:[a-zA-Z0-9]+]]
define void @assume_with_opbundles(i1 %true, ptr %p) {
  ; CHECK: %[[ALIGN:.+]] = llvm.mlir.constant(8 : i32) : i32
  ; CHECK:  llvm.intr.assume %[[TRUE]] ["align"(%[[PTR]], %[[ALIGN]] : !llvm.ptr, i32)] : i1
  call void @llvm.assume(i1 %true) ["align"(ptr %p, i32 8)]
  ret void
```
nit: Let's check the pointer as well.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:859]:
```suggestion
 size_t numOpBundleOperands = 0;
```

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:253]:
```suggestion
        auto bundleTag = cast<StringAttr>(std::get<2>(bundle)).getValue();
```

**Review by gysit — APPROVED** [review]
Nice!

LGTM modulo nit comments.

**Comment by Lancern** [comment]
> What are those burdens?

The main problem with properties is that they are not available through `Operation *`. You have to cast it to a `mlir::LLVM::FooOp` before getting/setting properties. However, when importing and exporting intrinsics, we all operate on `Operation *` and thus storing operand bundle tags in properties could introduce a bit trouble in such a case.

> Might also be worth summarizing super shortly in the commit message

Updated.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:884]: `bundleTag` is a `std::string` actually, don't miss the little `.str()` tail there :)

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Updated.

File: mlir/test/Target/LLVMIR/Import/intrinsic.ll
Comment (Lancern) [reply mlir/test/Target/LLVMIR/Import/intrinsic.ll]: Updated.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Updated.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp]: Updated.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:884]: Ah missed that one all fine then!

**Comment by Lancern** [comment]
Any other comments on this PR?

**Comment by gysit** [comment]
Looks good from my side! I believe you have now commit access? So feel free to land.

**Comment by Lancern** [comment]
Great, I'll merge now.

**Comment by Lancern** [comment]
Sorry but I'm not very familiar with how we deal with CI failures. Should I revert the patch and reland it after fixing? Or I just push another patch to fix it?

**Comment by gysit** [comment]
Yes the correct way is to revert the commit briefly mentioning the failure in the commit message (maybe just put a link to the build bot). And then reland later.

**Comment by Lancern** [comment]
Reverted, will land later after fixing the build error.

**Comment by gysit** [comment]
I guess you have to replace std::reduce since this is apparently not supported on power pc. std::accumulate seems to have uses in llvm so I supposes this maybe the way to go.

When you reland the PR it makes sense to add the commit hash of the original commit and of the revert commit to the commit message for transparency.

**Comment by Lancern** [comment]
Just re-landed. Let's see if CI build could pass now.

**Pull request merged**
